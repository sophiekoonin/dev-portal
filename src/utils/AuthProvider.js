import React from 'react'
import context from './AuthContext'

class AuthProvider extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      token: 'hello',
      user: null
    }
  }

  render() {
    const { Provider } = context
    return <Provider value={this.state}>{this.props.children}</Provider>
  }
}

export default AuthProvider
