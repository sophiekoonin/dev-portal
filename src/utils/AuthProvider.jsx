import React from 'react'
import context from './AuthContext'
import { apiBasePath } from '../config'
import Redirect from 'react-router/Redirect'

class AuthProvider extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      accessToken: null,
      username: null,
      hasError: false
    }
  }

  getAccessToken = async (username, password) => {
    await this.setState({ hasError: false })
    const res = await fetch(`${apiBasePath}/login`, {
      method: 'POST',
      headers: {
        'content-type': 'application/json'
      },
      body: JSON.stringify({ username, password })
    })
    const body = await res.json()
    if (res.status !== 200) {
      return this.setState({ hasError: true }, () => body['error'])
    } else {
      const { accessToken } = body
      return this.setState({ accessToken })
    }
  }

  render() {
    const { Provider } = context
    return (
      <Provider value={{ ...this.state, doLogin: this.getAccessToken }}>
        {this.props.children}
      </Provider>
    )
  }
}

export default AuthProvider
