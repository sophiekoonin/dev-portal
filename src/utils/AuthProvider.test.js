import React from 'react'
import AuthProvider from './AuthProvider'

describe('getAccessToken', () => {
  beforeEach(() => {
    fetch.resetMocks()
  })

  it('makes a POST request to /login with the given username and password', async () => {
    const wrapper = shallow(<AuthProvider />)

    fetch.mockResponseOnce(JSON.stringify({ accessToken: 'abc123' }))
    const instance = wrapper.instance()

    await instance.getAccessToken('myUsername', 'myPassword')
    expect(fetch.mock.calls.length).toBe(1)
    expect(fetch.mock.calls[0]).toMatchSnapshot()
  })

  it('sets state to hasError=true if the request returns a 401', async () => {
    const wrapper = shallow(<AuthProvider />)

    fetch.mockResponseOnce(JSON.stringify({ error: 'Bad credentials!' }), {
      status: 401
    })
    const instance = wrapper.instance()
    await instance.getAccessToken('myUsername', 'myPassword')
    expect(instance.state.hasError).toBe(true)
    expect(instance.state.accessToken).toBe(null)
  })

  it('stores access token in the state if the request returns a 200', async () => {
    const wrapper = shallow(<AuthProvider />)

    fetch.mockResponseOnce(JSON.stringify({ accessToken: 'abcde1234' }), {
      status: 200
    })
    const instance = wrapper.instance()
    await instance.getAccessToken('myUsername', 'myPassword')
    expect(instance.state.accessToken).toEqual('abcde1234')
  })
})
