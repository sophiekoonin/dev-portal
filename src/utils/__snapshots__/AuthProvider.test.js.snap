// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAccessToken makes a POST request to /login with the given username and password 1`] = `
Array [
  "https://guarded-thicket-22918.herokuapp.com/login",
  Object {
    "body": "{\\"username\\":\\"myUsername\\",\\"password\\":\\"myPassword\\"}",
    "headers": Object {
      "content-type": "application/json",
    },
    "method": "POST",
  },
]
`;
